Fix Bug 1513203, add Jest and True, and initial tests

schalkneethling opened the pull request:
@davidflanagan As mentioned in the bug, this adds two frameworks:

1. Jest
2. True

For True I had to also add Mocha as the test runner. I am looking into writing a little bridge so I can also run True with Jest, so that is temporary. I also added a small number of tests for each.

After this is merged, I will be pushing a lot more tests for both. r?

schalkneethling commented:
Looks like someone has already gotten Jest to work with True[https://github.com/oddbird/true/issues/132], looking into it. 🎉

davidflanagan commented on kuma/settings/common.py:
```diff
@@ -959,17 +959,17 @@ def pipeline_one_scss(slug, **kwargs):
         'source_filenames': (
-            'js/payments-handler.js',
-            'js/payments-faq.js',
+            'js/components/payments/payments-handler.js',
```
Don't you need to add the new payments-handler-utils.js file here?

davidflanagan replied:
Or maybe I just don't understand our js packaging scheme.

schalkneethling replied:
Nope, nope, you are correct. Doh!

schalkneethling commented on .eslintrc.js:
```diff
@@ -1,8 +1,15 @@
+    'parserOptions': {
+        'ecmaVersion': 6
+    },
+    "plugins": ["jest"],
```
> Or does the plugin enable some of its lint rules by default? Or maybe this is just so that you can have the jest/globals in env above.

It adds some of its own rules, but also adds the relevant globals so eslint does not complain about undefined globals such as `describe` and `it`

schalkneethling commented on package.json:
```diff
@@ -1,37 +1,43 @@
+        "optim:svg": "svgo jinja2/includes/icons/* --disable={removeViewBox,removeTitle} --multipass --pretty",
+        "precommit": "npm run eslint && npm run stylelint",
+        "test:js": "jest",
+        "test:sass": "mocha __sass_tests__/test-sass.js"
```
The colons is a convention. Just running `npm test` will not run them all but, you can add `npm-run-all` and then define `test` to run both. This just enable me to separate out the two sets so, you can either run JS tests or SaSS tests.

In general this is still evolving, this PR being the first step in getting the base in place.

schalkneethling replied:
This has been cleaned up. Now we only use Jest :tada:

schalkneethling commented on package.json:
```diff
@@ -1,37 +1,43 @@
+    "dependencies": {
+        "clean-css-cli": "4.2.1",
+        "eslint": "5.1.0",
+        "eslint-plugin-jest": "22.1.2",
```
You are 100% correct. I believe we also run eslint on CI so, adding the Jest plugin ensures eslint will not break because of Jest globals.

schalkneethling commented on kuma/settings/common.py:
```diff
@@ -959,17 +959,17 @@ def pipeline_one_scss(slug, **kwargs):
-            'js/payments-faq.js',
-            'js/payments-confirmation.js',
+            'js/components/payments/payments-faq.js',
+            'js/components/payments/payments-confirmation.js',
```
Moving it out of Python and into a Nodejs based tool chain is part of the work I/we need to do ;) I could totally have moved this is a separate PR, but as I was touching these, I thought I might as well start moving things into a more "component" based file structure at least.

schalkneethling replied:
Also, having tests will aid a great deal in the next steps with regards to the toolchain, as well as removing dependencies such as jQuery etc.

schalkneethling commented on jest.config.js:
```diff
@@ -0,0 +1,180 @@
+
+    // The glob patterns Jest uses to detect test files
+    // testMatch: [
+    //   "**/__tests__/**/*.js?(x)",
```
Good point. I was following the Jest convention over configuration path, but we surely can abandon that in favor of a probably cleaner folder structure.

schalkneethling commented on kuma/static/js/components/payments/payments-handler-utils.js:
```diff
@@ -0,0 +1,23 @@
+var mdn = window.mdn || {};
+var paymentsHandlerUtils = {
+    getNewValue: function(selectedAmount) {
```
Yup, pushed to  soon ;p

schalkneethling commented on kuma/static/js/components/payments/payments-handler-utils.js:
```diff
@@ -0,0 +1,23 @@
+    },
+    getSelectedAmount: function(value) {
+        'use strict';
+        return value % 1 === 0 ? parseInt(value, 10) : parseFloat(value).toFixed(2);
```
Also bothers me ;) I do not believe any math is done on this so, casting it to a `String` certainly makes sense.

schalkneethling replied:
It's a pitty that this was not documented when it was written. Currently it seems to me the only real reason `parseInt` is being used, is to convert something like `2.0` to `2`.

```
2.0 % 1 === 0 ? parseInt(2.0, 10) : parseFloat(2.0).toFixed(2);
// returns 2
```

Anything else like say 2.01, will go to the `parseFloat` side and be returned as a String

```
2.01 % 1 === 0 ? parseInt(2.0, 10) : parseFloat(2.0).toFixed(2);
// returns "2.01"
```

I guess what I am saying is, why not just parse all values to a `float` and return a String? @jwhitlock Do you know why this is?

/me sure I am missing something obvious.

davidflanagan replied:
I assume the reason not to use just the parseFloat() route is that then you'd end up with "$2.00" instead of "$2". But something like `parseFloat(value).toFixed(2).replace(/\.00$/, '') might do the same thing in an arguable cleaner way.

schalkneethling replied:
Yup. This was a decision made that I was not aware of so, I am not going to change it for now. Will put a pin in it, and cycle back later.

schalkneethling commented:
Thanks for the detailed review @davidflanagan - addressing your comments.

schalkneethling commented:
@davidflanagan Addressed all comments, also removed Mocha, so we are now only using Jest. When I say all comments, I mean all but https://github.com/mozilla/kuma/pull/5162#discussion_r240769278, which warrants more discussion.

schalkneethling commented:
Thanks @davidflanagan 🌮
